### ISSUE TITLE
IDE0270 creates invalid code

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#74460]
**Version Used**: 
Version 17.11.0 Preview 3.0 [35025.322.main]

**Steps to Reproduce**:
```C#
static void Repro(object? o)
{
    IList<int>? list = o as int[];
    if (list is null)
    {
        list = o as List<int>;
    }
}
``` 

**Diagnostic Id**:
IDE0270

**Expected Behavior**:
Diagnostic is not triggered if the resulting fixer's code won't compile.

**Actual Behavior**:
Diagnostic fires, and the fixer generates this code:
```C#
static void Repro(object? o)
{
    IList<int>? list = o as int[] ?? o as List<int>;
}
```
which fails to compile with:
```
error CS0019: Operator '??' cannot be applied to operands of type 'int[]' and 'List<int>'
```

**Comment by CyrusNajmabadi** [comment]
If someone wants to fix this, two options are possible.  Either do not offer the suggestion here.  Or offer, but insert an explicit cast to teh destination type.

**Comment by Orachor** [comment]
I'll try taking a look at this :)
